### PR TITLE
#31262 add a public hasher function for HashSet and HashMap

### PR DESCRIPTION
**Pull request by rthomas**
add a public hasher function for HashSet and HashMap

### CODE REVIEW

File: src/libstd/collections/hash/set.rs [hunk 1]
```diff
@@ -193,6 +193,14 @@ impl<T, S> HashSet<T, S>
         }
     }
 
+    /// Returns the hasher that is used or the detault hasher if none is
+    /// specified.
+    #[inline]
+    #[stable(feature = "hashmap_build_hasher", since = "1.7.0")]
```
Comment (apasel422) [inline src/libstd/collections/hash/set.rs]: Why is this stable, but the corresponding method for `HashMap` unstable?

File: src/libstd/collections/hash/map.rs [hunk 2]
```diff
@@ -642,6 +642,15 @@ impl<K, V, S> HashMap<K, V, S>
         HashMap::with_capacity_and_hasher(capacity, hash_state)
     }
 
+    /// Returns the hasher that is used or the detault hasher if none is
```
Comment (apasel422) [inline src/libstd/collections/hash/map.rs]: s/detault/default/

File: src/libstd/collections/hash/set.rs [hunk 3]
```diff
@@ -193,6 +193,14 @@ impl<T, S> HashSet<T, S>
         }
     }
 
+    /// Returns the hasher that is used or the detault hasher if none is
```
Comment (apasel422) [inline src/libstd/collections/hash/set.rs]: s/detault/default/

File: src/libstd/collections/hash/set.rs (hunk 1 above)
Comment (rthomas) [inline src/libstd/collections/hash/set.rs]: Sorry, should have been stable. Fixed.

File: src/libstd/collections/hash/map.rs (hunk 2 above)
Comment (rthomas) [inline src/libstd/collections/hash/map.rs]: Done.

File: src/libstd/collections/hash/set.rs (hunk 3 above)
Comment (rthomas) [inline src/libstd/collections/hash/set.rs]: Done.

File: src/libstd/collections/hash/map.rs [hunk 4]
```diff
@@ -642,6 +642,14 @@ impl<K, V, S> HashMap<K, V, S>
         HashMap::with_capacity_and_hasher(capacity, hash_state)
     }
 
+    /// Returns the hasher that is used or the default hasher if none is
```
Comment (apasel422) [inline src/libstd/collections/hash/map.rs]: I think this would be better as something like:

```
Returns a reference to the map's hasher.
```

Comment (rthomas) [inline src/libstd/collections/hash/map.rs]: Done.

**Comment by apasel422** [comment]
CC @rust-lang/libs

File: src/libstd/collections/hash/map.rs [hunk 5]
```diff
@@ -642,6 +642,13 @@ impl<K, V, S> HashMap<K, V, S>
         HashMap::with_capacity_and_hasher(capacity, hash_state)
     }
 
+    /// Returns a reference to the maps hasher.
```
Comment (Gankro) [inline src/libstd/collections/hash/map.rs]: `map's`

File: src/libstd/collections/hash/set.rs [hunk 6]
```diff
@@ -193,6 +193,13 @@ impl<T, S> HashSet<T, S>
         }
     }
 
+    /// Returns a reference to the underlying HashMaps hasher.
```
Comment (Gankro) [inline src/libstd/collections/hash/set.rs]: I would just change this to "the set's hasher"; the fact that sets are "just maps" is an impl detail.

**Comment by Gankro** [comment]
I don't believe we want to make this functionality insta-stable? Probably want to introduce a new feature for this

**Comment by Gankro** [comment]
CC https://github.com/rust-lang/rust/issues/31262

File: src/libstd/collections/hash/map.rs (hunk 5 above)
Comment (rthomas) [inline src/libstd/collections/hash/map.rs]: Fixed.

File: src/libstd/collections/hash/set.rs (hunk 6 above)
Comment (rthomas) [inline src/libstd/collections/hash/set.rs]: Fixed.

**Comment by pnkfelix** [comment]
> I don't believe we want to make this functionality insta-stable? Probably want to introduce a new feature for this.

I second this,

**Comment by rthomas** [comment]
@pnkfelix yep - just updated the issue and will push the PR when tests complete.

**Comment by apasel422** [comment]
@rthomas Thanks! Could you squash these commits?

**Comment by alexcrichton** [comment]
The libs team discussed this during triage today and the conclusion was that this is good to go. As well as squashing like @apasel422 mentioned, could you also remove the `#[inline]` tags as well? These are already inlined across crates because of the generics and probably don't really need the extra help from LLVM.

Other than that looks good to me, thanks for the PR @rthomas!

**Comment by rthomas** [comment]
Thanks @apasel422 and @alexcrichton, I've removed the `#[inline]` tags and squashed the commits.

**Pull request merged**